import os, sys
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import pandas as pd

#use os.chdir(mydir) to change to the directory that contains your gage results and your lengths file

filename1 = 'diso1_a2_pass_fq_read_lengths.txt'
filename2 = 'iso1_onp_gage_results.csv'

myReadLengthsPD = pd.read_csv(filename1, sep='\t', header=None)
myReadLengthsPD.columns = ['Read Name', 'Read Lengths']

myArraygte50kPD = myReadLengthsPD.ix[ myReadLengthsPD.iloc[:,1] >= 50000,1]
myArraylt50kPD = myReadLengthsPD.ix[ myReadLengthsPD.iloc[:,1] < 50000,1]

sns.set(font_scale=2)
sns.set_style('ticks')
#sns.set_palette('colorblind')
sns.set_palette('pastel')
ax = sns.distplot(myArraygte50kPD, kde=False)
ax.set_title('D. melanogaster ISO1 ONT Reads >= 50Kb Distribution')
sns.despine()

ax = sns.distplot(myArraylt50kPD, kde=False)
ax.set_title('D. melanogaster ISO1 ONT Reads < 50Kb Distribution')
sns.despine()
#ax = ''

sns.set_style("whitegrid")
ax = sns.distplot(myReadLengthsPD.ix[:,1], kde_kws=dict(cumulative=True))
ax.invert_xaxis()
ax.set_title('D. melanogaster ISO1 ONT Reads CDF')
plt.yticks([0.5,0.75,0.95,1.0])
sns.despine()

myGageResults = pd.read_csv(filename2)

#canu only
#myGageResults.ix[0,['asms', 'SNPs','Indels < 5bp', 'Indels >= 5', 'Inversions', 'Relocation', 'Translocation']]
#all
#myGageResults.loc[:,['asms', 'SNPs','Indels < 5bp', 'Indels >= 5', 'Inversions', 'Relocation', 'Translocation']]

sns.set(font_scale=2)
sns.set_style('ticks')
sns.set_palette('colorblind')

plt.subplot(231)
sns.barplot(myGageResults.loc[:,'asms', ], myGageResults.loc[:,'SNPs'])
plt.xlabel('')
plt.subplot(232)
sns.barplot(myGageResults.loc[:,'asms', ], myGageResults.loc[:,'Indels < 5bp'])
plt.xlabel('')
plt.subplot(233)
sns.barplot(myGageResults.loc[:,'asms', ], myGageResults.loc[:,'Indels >= 5'])
plt.xlabel('')
plt.subplot(234)
sns.barplot(myGageResults.loc[:,'asms', ], myGageResults.loc[:,'Inversions'])
plt.xlabel('')
plt.subplot(235)
sns.barplot(myGageResults.loc[:,'asms', ], myGageResults.loc[:,'Relocation'])
plt.xlabel('Assembly Type')
plt.subplot(236)
sns.barplot(myGageResults.loc[:,'asms', ], myGageResults.loc[:,'Translocation'])
plt.xlabel('')
sns.despine()
plt.suptitle('Genome Assembly GAGE Structural Discordance')
plt.subplots_adjust(top=0.90, bottom=0.10, left=0.05, right=0.95, hspace=0.20, wspace=0.25)

sns.set(font_scale=2)
sns.set_style('ticks')
sns.set_palette('colorblind')
plt.subplot(241)
sns.barplot(myGageResults.loc[:,'asms',], myGageResults.loc[:,'Misassemblies'])
plt.xlabel('')
plt.subplot(242)
sns.barplot(myGageResults.loc[:,'asms',], myGageResults.loc[:,'Misassembled Contigs'])
plt.xlabel('')
plt.subplot(243)
sns.barplot(myGageResults.loc[:,'asms',], myGageResults.loc[:,'Mismatches per 100 Kbp'])
plt.xlabel('')
plt.subplot(244)
sns.barplot(myGageResults.loc[:,'asms',], myGageResults.loc[:,'Indels per 100 Kbp'])
plt.xlabel('')
plt.subplot(245)
sns.barplot(myGageResults.loc[:,'asms',], myGageResults.loc[:,'Local Misassemblies'])
plt.xlabel('')
plt.subplot(246)
sns.barplot(myGageResults.loc[:,'asms',], myGageResults.loc[:,'Whole Unaligned Contigs'])
plt.xlabel('')
plt.subplot(247)
sns.barplot(myGageResults.loc[:,'asms',], myGageResults.loc[:,'Unaligned Length'])
plt.xlabel('')
plt.subplot(248)
sns.barplot(myGageResults.loc[:,'asms',], myGageResults.loc[:,'Genome fraction (%)'])
plt.yticks(np.arange(80.0, 101.0, 2.0))
plt.ylim(ymin=80,ymax=100)
plt.xlabel('')
sns.despine()
plt.suptitle('Genome Assembly GAGE Statistics')
plt.subplots_adjust(top=0.90, bottom=0.10, left=0.05, right=0.95, hspace=0.20, wspace=0.25)

sns.set(font_scale=2)
sns.set_style('ticks')
sns.set_palette('colorblind')
plt.subplot(4,3,1)
sns.barplot(myGageResults.loc[:,'asms',], myGageResults.loc[:,'Genome fraction (%)'])
plt.yticks(np.arange(80.0, 101.0, 2.0))
plt.ylim(ymin=80,ymax=100)
plt.xlabel('')
plt.subplot(4,3,2)
sns.barplot(myGageResults.loc[:,'asms',], myGageResults.loc[:,'NGA50'])
plt.xlabel('')
plt.subplot(4,3,3)
sns.barplot(myGageResults.loc[:,'asms',], myGageResults.loc[:,'Unaligned Length'])
plt.xlabel('')
plt.subplot(4,3,4)
sns.barplot(myGageResults.loc[:,'asms',], myGageResults.loc[:,'Misassembled Contigs'])
plt.xlabel('')
plt.subplot(4,3,5)
sns.barplot(myGageResults.loc[:,'asms',], myGageResults.loc[:,'Misassemblies'])
plt.xlabel('')
plt.subplot(4,3,6)
sns.barplot(myGageResults.loc[:,'asms',], myGageResults.loc[:,'Local Misassemblies'])
plt.xlabel('')
plt.subplot(4,3,7)
sns.barplot(myGageResults.loc[:,'asms', ], myGageResults.loc[:,'Relocation'])
plt.xlabel('')
plt.subplot(4,3,8)
sns.barplot(myGageResults.loc[:,'asms', ], myGageResults.loc[:,'Inversions'])
plt.xlabel('')
plt.subplot(4,3,9)
sns.barplot(myGageResults.loc[:,'asms', ], myGageResults.loc[:,'Translocation'])
plt.xlabel('')
plt.subplot(4,3,10)
sns.barplot(myGageResults.loc[:,'asms', ], myGageResults.loc[:,'SNPs'])
plt.xlabel('')
plt.subplot(4,3,11)
sns.barplot(myGageResults.loc[:,'asms', ], myGageResults.loc[:,'Indels < 5bp'])
plt.xlabel('')
plt.subplot(4,3,12)
sns.barplot(myGageResults.loc[:,'asms', ], myGageResults.loc[:,'Indels >= 5'])
plt.xlabel('')
sns.despine()
plt.suptitle('Genome Assembly GAGE Structural Discordance')
plt.subplots_adjust(top=0.90, bottom=0.05, left=0.05, right=0.95, hspace=0.20, wspace=0.25)

sns.set(font_scale=2)
sns.set_style('ticks')
sns.set_palette('colorblind')
plt.subplot(4,3,1)
sns.barplot(myGageResults.loc[0:3,'asms',], myGageResults.loc[:,'Genome fraction (%)'])
plt.yticks(np.arange(80.0, 101.0, 2.0))
plt.ylim(ymin=80,ymax=100)
plt.xlabel('')
plt.subplot(4,3,2)
sns.barplot(myGageResults.loc[0:3,'asms',], myGageResults.loc[:,'NGA50'])
plt.xlabel('')
plt.subplot(4,3,3)
sns.barplot(myGageResults.loc[0:3,'asms',], myGageResults.loc[:,'Unaligned Length'])
plt.xlabel('')
plt.subplot(4,3,4)
sns.barplot(myGageResults.loc[0:3,'asms',], myGageResults.loc[:,'Misassembled Contigs'])
plt.xlabel('')
plt.subplot(4,3,5)
sns.barplot(myGageResults.loc[0:3,'asms',], myGageResults.loc[:,'Misassemblies'])
plt.xlabel('')
plt.subplot(4,3,6)
sns.barplot(myGageResults.loc[0:3,'asms',], myGageResults.loc[:,'Local Misassemblies'])
plt.xlabel('')
plt.subplot(4,3,7)
sns.barplot(myGageResults.loc[0:3,'asms', ], myGageResults.loc[:,'Relocation'])
plt.xlabel('')
plt.subplot(4,3,8)
sns.barplot(myGageResults.loc[0:3,'asms', ], myGageResults.loc[:,'Inversions'])
plt.xlabel('')
plt.subplot(4,3,9)
sns.barplot(myGageResults.loc[0:3,'asms', ], myGageResults.loc[:,'Translocation'])
plt.xlabel('')
plt.subplot(4,3,10)
sns.barplot(myGageResults.loc[0:3,'asms', ], myGageResults.loc[:,'SNPs'])
plt.xlabel('')
plt.subplot(4,3,11)
sns.barplot(myGageResults.loc[0:3,'asms', ], myGageResults.loc[:,'Indels < 5bp'])
plt.xlabel('')
plt.subplot(4,3,12)
sns.barplot(myGageResults.loc[0:3,'asms', ], myGageResults.loc[:,'Indels >= 5'])
plt.xlabel('')
sns.despine()
plt.suptitle('Genome Assembly GAGE Structural Discordance')
plt.subplots_adjust(top=0.90, bottom=0.05, left=0.05, right=0.95, hspace=0.20, wspace=0.25)
